### PR TITLE
*: support history read.

### PR DESCRIPTION
**Pull request by coocood**
This commit only handles the case when scheme does not change.
Use history schema will be supported in the following PR.

### CODE REVIEW

**Comment by coocood** [comment]
@shenli @zimulala PTAL

File: context/context.go [hunk 1]
```diff
@@ -24,6 +24,9 @@ type Context interface {
 	// GetTxn gets a transaction for further execution.
 	GetTxn(forceNew bool) (kv.Transaction, error)
 
+	// GetClient gets a kv.Client
```
Comment (zimulala) [inline context/context.go]: Add . at the end of a line.

File: sessionctx/variable/session.go [hunk 2]
```diff
@@ -107,6 +110,9 @@ type SessionVars struct {
 
 	// InUpdateStmt indicates if the session is handling update stmt.
 	InUpdateStmt bool
+
+	// SnapshotTS is used for reading history data. for simplicity, SnapshotTS only supports distsql request,
```
Comment (zimulala) [inline sessionctx/variable/session.go]: Use . as a statement terminator.

File: sessionctx/variable/session.go [hunk 3]
```diff
@@ -162,6 +168,12 @@ func GetCharsetInfo(ctx context.Context) (charset, collation string) {
 	return
 }
 
+// GetSnapshotTS gets snapshot timestamp that has been set by set variable statement.
+func GetSnapshotTS(ctx context.Context) uint64 {
+	sessionVars := GetSessionVars(ctx)
```
Comment (hanfei1991) [inline sessionctx/variable/session.go]: Will sessionVars be null pointer?

File: sessionctx/variable/session.go [hunk 4]
```diff
@@ -226,11 +238,32 @@ func (s *SessionVars) SetSystemVar(key string, value types.Datum) error {
 		} else {
 			s.StrictSQLMode = false
 		}
+	} else if key == "tidb_snapshot" {
```
Comment (hanfei1991) [inline sessionctx/variable/session.go]: I think it's better to set "tidb_snapshot" as a const string .

File: sessionctx/variable/session.go [hunk 5]
```diff
@@ -226,11 +238,32 @@ func (s *SessionVars) SetSystemVar(key string, value types.Datum) error {
...
+		err = s.setSnapshotTS(sVal)
+		if err != nil {
+			return errors.Trace(err)
+		}
 	}
 	s.systems[key] = sVal
 	return nil
 }
 
+const epochShiftBits = 18
```
Comment (hanfei1991) [inline sessionctx/variable/session.go]: Add comment for this global var.

File: executor/builder.go [hunk 6]
```diff
@@ -606,13 +601,19 @@ func (b *executorBuilder) buildTableDual(v *plan.TableDual) Executor {
...
+	var txn kv.Transaction
+	var err error
+	startTS := variable.GetSnapshotTS(b.ctx)
+	if startTS == 0 {
+		txn, err = b.ctx.GetTxn(false)
+		if err != nil {
+			b.err = errors.Trace(err)
+			return nil
+		}
+		startTS = txn.StartTS()
```
Comment (zimulala) [inline executor/builder.go]: line604-613 can wrap a function.

File: sessionctx/variable/session.go (hunk 3 above)
Comment (coocood) [inline sessionctx/variable/session.go]: sessionVars is set in `CreateSession`, so it will never be nil.

**Comment by zimulala** [comment]
LGTM

**Comment by coocood** [comment]
@hanfei1991 PTAL

File: sessionctx/variable/sysvar.go [hunk 7]
```diff
@@ -580,6 +580,7 @@ var defaultSysVars = []*SysVar{
 	{ScopeGlobal | ScopeSession, "min_examined_row_limit", "0"},
 	{ScopeGlobal, "sync_frm", "ON"},
 	{ScopeGlobal, "innodb_online_alter_log_max_size", "134217728"},
+	{ScopeSession, "tidb_snapshot", ""},
```
Comment (hanfei1991) [inline sessionctx/variable/sysvar.go]: this should use tidbSnapshot const

Comment (coocood) [inline sessionctx/variable/sysvar.go]: This involves too many strings, should not change it in this PR.

Comment (hanfei1991) [inline sessionctx/variable/sysvar.go]: but you have defined it....

**Comment by hanfei1991** [comment]
LGTM

File: executor/adapter.go [hunk 8]
```diff
@@ -96,6 +97,15 @@ func (a *statement) Exec(ctx context.Context) (ast.RecordSet, error) {
 	}
 
 	if len(e.Fields()) == 0 && len(e.Schema()) == 0 {
+		// Write statements do not have record set, check if snapshot ts is set.
+		switch e.(type) {
+		case *DeleteExec, *InsertExec, *UpdateExec, *LoadData, *DDLExec:
+			snapshotTS := variable.GetSnapshotTS(ctx)
+			if snapshotTS != 0 {
```
Comment (shenli) [inline executor/adapter.go]: How about if snapshotTS != 0 & ok := e.(*SelectStmt); !ok { return error}?

Comment (hanfei1991) [inline executor/adapter.go]: UnionStmt is also valid.

Comment (coocood) [inline executor/adapter.go]: We need to do `Set`, `Show` or some other statements.

**Comment by shenli** [comment]
LGTM

**Pull request merged**